pkg/controller/job/job_controller_test.go

```diff
@@ -1338,3 +1339,70 @@ func TestJobBackoffReset(t *testing.T) {
 		}
 	}
 }
+
+var _ workqueue.RateLimitingInterface = &fakeRateLimitingQueue{}
+
+type fakeRateLimitingQueue struct {
+	workqueue.Interface
+	requeues int
+	item     interface{}
+	duration time.Duration
+}
+
+func (f *fakeRateLimitingQueue) AddRateLimited(item interface{}) {}
+func (f *fakeRateLimitingQueue) Forget(item interface{})         {}
+func (f *fakeRateLimitingQueue) NumRequeues(item interface{}) int {
+	return f.requeues
+}
+func (f *fakeRateLimitingQueue) AddAfter(item interface{}, duration time.Duration) {
+	f.item = item
+	f.duration = duration
+}
+
+func TestJobBackoff(t *testing.T) {
+	job := newJob(1, 1, 1)
+	oldPod := newPod(fmt.Sprintf("pod-%v", rand.String(10)), job)
+	oldPod.Status.Phase = v1.PodRunning
+	oldPod.ResourceVersion = "1"
+	newPod := oldPod.DeepCopy()
+	newPod.ResourceVersion = "2"
+
+	testCases := map[string]struct {
+		// inputs
+		requeues int
+		phase    v1.PodPhase
+
+		// expectation
+		backoff int64
+	}{
+		"1st failure": {0, v1.PodFailed, int64(0) * DefaultJobBackOff.Nanoseconds()},
+		"2nd failure": {1, v1.PodFailed, int64(1) * DefaultJobBackOff.Nanoseconds()},
+		"3rd failure": {2, v1.PodFailed, int64(2) * DefaultJobBackOff.Nanoseconds()},
+		"1st success": {0, v1.PodSucceeded, int64(0) * DefaultJobBackOff.Nanoseconds()},
+		"2nd success": {1, v1.PodSucceeded, int64(0) * DefaultJobBackOff.Nanoseconds()},
+		"1st running": {0, v1.PodSucceeded, int64(0) * DefaultJobBackOff.Nanoseconds()},
+		"2nd running": {1, v1.PodSucceeded, int64(0) * DefaultJobBackOff.Nanoseconds()},
```

Comment:
running cases and success cases are the same?

Reply:
Yes they are, but I've decided to put them for clarity :-)